src/web100srv.c

```diff
@@ -1527,647 +1326,364 @@ void child_process(int parent_pipe, SSL_CTX *ssl_context, int ctlsockfd) {
+  // require modification to allow more credentials to be created/passed
+  // between the user and this application.
+
+  child_pid = fork();
+  if (child_pid == -1) {
+    // an error occured, log it and return
+    log_println(0, "fork() failed, errno = %d (%s)", errno, strerror(errno));
+    close(child_pipe[0]);
+    close(child_pipe[1]);
+    close(ctlsockfd);
+    free(new_child);
+    return NULL;
+  } else if (child_pid == 0) {
+    // This is the child. Clean up and close the resources the child does not
+    // need, then call the child_process routine which will initialize the
+    // connection and then wait for the go/queue/nogo signals from the parent
+    // process.
+    log_println(4, "Child thinks pipe() returned fd0=%d, fd1=%d for pid=%d",
+                child_pipe[0], child_pipe[1], child_pid);
+    close(listenfd);
+    close(child_pipe[1]);
+    child_process(child_pipe[0], ssl_context, ctlsockfd);
+    log_println(1, "The child returned! This should never happen.");
+    exit(1);
+  }
 
-    if ((multiple == 1) && (mclients > waiting)) {
-      log_println(
-          5,
-          "Multi-client mode has uncaught terminated clients mclient=%d, "
-          "waiting=%d",
-          mclients, waiting);
-      while (mclients > waiting) {
-        child_sig(0);
-        mclients--;
-        log_println(5, "Removed terminated client from count mclients=%d",
-                    mclients);
-      }
-    }
+  // this is the parent process, handle scheduling and
+  // queuing of multiple incoming clients
+  log_println(5, "Parent process spawned child = %d", child_pid);
+  log_println(5, "Parent thinks pipe() returned fd0=%d, fd1=%d", child_pipe[0],
+              child_pipe[1]);
+
+  // Close the open resources that should only be used by the child.
+  close(child_pipe[0]);
+  close(ctlsockfd);
+
+  // Initialize the members of new_child
+  new_child->pid = child_pid;
+  strlcpy(new_child->addr, rmt_addr, sizeof(new_child->addr));
+  strlcpy(new_child->host, rmt_host, sizeof(new_child->host));
+  new_child->qtime = time(0);
+  new_child->pipe = child_pipe[1];
+  new_child->running = 0;
+  new_child->next = NULL;
+
+  return new_child;
+}
 
-    FD_ZERO(&rfd);
-    FD_SET(listenfd, &rfd);
-    if (waiting > 0) {
-      // there are clients waiting
-      sel_tv.tv_sec = 3;  // 3 seconds == WAIT_TIME_SRVR
-      sel_tv.tv_usec = 0;
-      log_println(3, "Waiting for new connection, timer running");
-      do {
-        retcode = select(listenfd + 1, &rfd, NULL, NULL, &sel_tv);
-        // retry if a signal caused select() to exit.
-      } while ((retcode == -1) && (errno == EINTR));
-      tt = time(0);
+/**
+ * Enqueues a new ndtchild struct onto the queue of clients to be run.  Will
+ * not enqueue the child unless the server has spare capacity.  This function
+ * takes ownership of the memory pointed to by new_child, and either frees that
+ * memory or passes ownership to the linked list that goes from head to tail.
```

Comment:
>the linked list that goes from head to tail

I don't understand what this means. Don't all linked lists go from head to tail?